### PR TITLE
code update to pass flake8 rule C416

### LINKED ISSUE #24935: flake8 scans failing on CI/CD pipelines
CI/CD jobs in many PRs are failing when scanning the code with flake8.

The issue seems to break on the base code, not on PR-ed commits. Are flake8 rules pinned?
```
sympy/tensor/array/expressions/from_array_to_matrix.py:874:16: C416 Unnecessary dict comprehension - rewrite using dict().
Error: Process completed with exit code 1.
```

See
https://github.com/sympy/sympy/actions/runs/4456494418/jobs/7826977213?pr=24253#step:7:1
https://github.com/sympy/sympy/actions/runs/4456527846/jobs/7827030982
https://github.com/sympy/sympy/actions/runs/4456328202/jobs/7826700508

### PR DESCRIPTION
**Pull request by maciejskorski**
#### References to other Issues or PRs

Fixes #24935

#### Brief description of what is fixed or changed

*Remove unnecessary uses of dict comprehensions* as [demanded by flake8 rule C416](https://pypi.org/project/flake8-comprehensions/), specifically:
```
sympy/parsing/mathematica.py:388:17: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/physics/units/dimensions.py:442:16: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/polys/multivariate_resultants.py:130:28: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/printing/tests/test_tensorflow.py:57:19: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/printing/tests/test_tensorflow.py:89:19: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/printing/tests/test_tensorflow.py:117:19: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/printing/tests/test_tensorflow.py:134:19: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/printing/tests/test_tensorflow.py:150:19: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/solvers/ode/ode.py:2162:17: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/solvers/ode/riccati.py:274:12: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/solvers/ode/systems.py:1885:21: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/stats/rv.py:1308:26: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/stats/rv.py:1312:35: C416 Unnecessary dict comprehension - rewrite using dict().
sympy/tensor/array/expressions/from_array_to_matrix.py:874:16: C416 Unnecessary dict comprehension - rewrite using dict().
```
#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by moorepants** [comment]
This seems like an unusual change. The "old" way to do this is `dict(zip())` and dictionary comprehensions were introduced to python sometime in Python 3 I believe, so you are changing the new way back to the old way. I think we should allow people to write this in either way.

**Comment by maciejskorski** [comment]
> This seems like an unusual change. The "old" way to do this is `dict(zip())` and dictionary comprehensions were introduced to python sometime in Python 3 I believe, so you are changing the new way back to the old way. I think we should allow people to write this in either way.

Thanks for the input. Note that this is not my proposal, but merely aligning with the linter (somebody has set up this rule) and supported by the code owner (see the comment by @oscarbenjamin in the issue). I also see the advantage of fixing the failing jobs immediately.

**Comment by moorepants** [comment]
> Personally I agree with you and would love to maintain that freedom, but like said above the decisions seem to have been already made.

Decisions are made on each pull request, in general.

**Comment by maciejskorski** [comment]
> Decisions are made on each pull request, in general.

Perhaps it is worth explaining that this not limits the use of comprehensions, but only *optimizes unnecessary comprehensions*. You should still use comprehensions where it makes sense, in particular these two patterns: dictionary reversions and dictionary processing.

**Comment by moorepants** [comment]
Both versions are valid Python and it is very unlikely that this rule will improve performance in SymPy, so it seems unnecessary linting. I see now that this is a hack to fix the failing tests, but isn't the correct fix to setup pytest in sympy so that it is impervious to upstream changes in the library. Is a pin on pytest the actual simple fix?

**Comment by Skylion007** [comment]
> Both versions are valid Python and it is very unlikely that this rule will improve performance in SymPy, so it seems unnecessary linting. I see now that this is a hack to fix the failing tests, but isn't the correct fix to setup pytest in sympy so that it is impervious to upstream changes in the library. Is a pin on pytest the actual simple fix?

It will generate less Python bytecode since it won't have to unpack/repack the tuple for the k,v pair for the trivial unnecessary noop comprehension. C416 was added because it is slightly more efficient.

Whether flake8 and flake8 plugins should be pinned is a separate issue to this PR.

**Comment by maciejskorski** [comment]
> Both versions are valid Python and it is very unlikely that this rule will improve performance in SymPy, so it seems unnecessary linting. I see now that this is a hack to fix the failing tests, but isn't the correct fix to setup pytest in sympy so that it is impervious to upstream changes in the library. Is a pin on pytest the actual simple fix?

Given that I volunteered and did the job as suggested by @oscarbenjamin what I am bringing is the simplest fix IMHO.

Somebody reviewed the linter update, somebody approved it, somebody suggested going ahead along this change. I am not here to redesign, but to volunteer.

**Review by Skylion007 — APPROVED** [review]

**Comment by moorepants** [comment]
There seems to be a new tendency and desire to introduce linting in SymPy because we can and because it is an easy PR to contribute. We should add linting rules if it solves a real issue or that we really want to enforce a specific style. This linting rule was created to solve performance issues (in general) but there is no demonstration that this solves a performance issue in SymPy. It happens to also solve our CI failing issue, but that is just today. What happens when pytest fails for another new upstream change? I don't think we should start the pattern of introducing linting rules because upstream changes occur in the testing tools. This solution just seems like a bad precedent to set.

**Comment by maciejskorski** [comment]
> There seems to be a new tendency and desire to introduce linting in SymPy because we can and because it is an easy PR to contribute. We should add linting rules if it solves a real issue or that we really want to enforce a specific style. This linting rule was created to solve performance issues (in general) but there is no demonstration that this solves a performance issue in SymPy. It happens to also solve our CI failing issue, but that is just today. What happens when pytest fails for another new upstream change? I don't think we should start the pattern of introducing linting rules because upstream changes occur in the testing tools. This solution just seems like a bad precedent to set.

When opening the issue, I also suggested pinning versions. Here we are suffering from the classical problem of *transitive dependencies*: sympy calls flake8 which calls its child dependencies that update overnight. I fully support pinning as a solid solution, but also agree with the linter on this specific case.

**Comment by oscarbenjamin** [comment]
> There seems to be a new tendency and desire to introduce linting in SymPy because we can and because it is an easy PR to contribute.

No, that is not what is happening here. CI is broken because of an update in one of the linters and @maciejskorski is trying to fix it which should be commended.

The code changes are an improvement. We can separately discuss whether or not we want to use any specific linting rules but right now this PR is good so I'm going to merge it.

**Comment by oscarbenjamin** [comment]
Thanks @maciejskorski !

**Comment by moorepants** [comment]
@oscarbenjamin, I believe this is the first time in my 14 years of working in the SymPy development team that I've seen someone merge a PR when there is an outstanding -1. That is quite concerning to me. As I understand, we operate on a tacit consent model for merging PRs, meaning that it is up to the author of the PR to work to get consensus if any issues are raised. That's exactly what we did in the recently merged switch to pytest. Given that you are a large contributor to SymPy with commit rights, I would hope that you are acutely aware of that.

**Comment by oscarbenjamin** [comment]
I don't see any -1 above. My reading of the above is that you are suggesting that the flake8 rule should not be enforced but this PR does not change any linting rules.

Do you object to the actual changes in this PR?

This PR is removing redundant comprehensions. you don't need to write:
```
stuff = [x for x in things]
```
because it is more efficient and simpler to write
```
stuff = list(things)
```
There is never any reason to use a list comprehension just to turn an iterable into a list.

This PR just does the equivalent for redundant dict comprehensions. A dict comprehension is useful if you are changing the values or keys somehow or filtering them. If you are just putting them into a dict then the comprehension is both unnecessary and less efficient.

**Comment by moorepants** [comment]
I do not consent to adopting the linting rule from the upstream update that triggered this PR and I think that it is pretty clear from my above statements that I was raising issue with that.

**Comment by oscarbenjamin** [comment]
> I do not consent to adopting the linting rule from the upstream update that triggered this PR and I think that it is pretty clear from my above statements that I was raising issue with that.

That is how I understood it. The linting rule is a separate question though from wanting the actual change here which is both a code improvement and an immediate fix for CI. I merged this because by my reading you did not object to the actual changes in this PR but of course we can reconsider that.

There are several options that we can consider in relation to the linting rule(s):

- We can pin the version of flake8-comprehensions.
- We can selectively disable this error code in the config.
- We can change the config to explicitly list the error codes that are wanted.

Version pinning isn't good because it always gets out of date and is work to maintain. Selectively disabling this error code leaves open the possibility that a future error code could be added leading to a similar situation of broken CI. 

If flake8-comprehensions is going to add new error codes like this then probably the best way is to explicitly list only the error codes that are wanted and are known to pass. The current list of error codes is here:
https://pypi.org/project/flake8-comprehensions/

The current filter is:
https://github.com/sympy/sympy/blob/9b7d671ea15131aab0fe80c8cd73ed8d0cca7be5/.flake8#L5-L6
That selects all rules that begin with `C4`. We can add ignore `C416` or we can change the selection to an explicit list of error codes or we can make the filter more selective. The corresponding ruff config is a little more selective and so  think would not pick up on this new rule:
https://github.com/sympy/sympy/blob/9b7d671ea15131aab0fe80c8cd73ed8d0cca7be5/pyproject.toml#L27-L30

**Comment by moorepants** [comment]
> That is how I understood it. The linting rule is a separate question though from wanting the actual change here which is both a code improvement and an immediate fix for CI.

That is your opinion, which you have the right to. Merging the PR in an autocratic way because your opinion doesn't align with others in the discussion is the concern I'm now raising. The technical details are irrelevant here at this point, given that you merged the PR. The appropriate behavior is to raise your points before the merge.

**Comment by oscarbenjamin** [comment]
Okay I'm sorry Jason. I understood your objection as being about the linting rule rather than the content of the PR itself and I wanted to get CI fixed so I guess I rushed merging this. We can of course consider reverting this if you are unhappy with the specific changes here.

Either way we should follow up by considering what we want from this linter and we should change the config to be more selective in some way so that new lint codes don't appear unexpectedly like this.

File: sympy/physics/units/dimensions.py
Comment (oscarbenjamin) [inline sympy/physics/units/dimensions.py:442]: I wonder what the purpose of this lne is. If `dimdep` is a dict then is there really any reason we can't just return it? At least if we want a copy then that's just `dict(dimdep)` or `dimdep.copy()`.

Comment (oscarbenjamin) [reply sympy/physics/units/dimensions.py:442]: I suspect that this kind of things comes from continuous edits to the code that end up with something sloppy that didn't really need to be there. This is probably the sort of thing where a linter rule (if enabled when the code was last edited) would have prompted the author to simplify this code.

Comment (maciejskorski) [reply sympy/physics/units/dimensions.py:442]: Don't know without a hint/strong typing I am afraid. Could be a modified class or so. 
Hence, I chose the safe solution.

Comment (maciejskorski) [reply sympy/physics/units/dimensions.py:442]: > This is probably the sort of thing where a linter rule (if enabled when the code was last edited) would have prompted the author to simplify this code.

Which is why I share your point of view that exploring the linter - rather than silencing it by pinning to old versions - could gradually improve legacy issues.

**Comment by maciejskorski** [comment]
> I do not consent to adopting the linting rule from the upstream update that triggered this PR and I think that it is pretty clear from my above statements that I was raising issue with that.

As the PR author I ask for feedback but need clear answers. I am afraid that you were not fully clear on what you object. 
The linting rule has been approved elsewhere, which is why I consider it a legitimate starting point. 

Please keep in mind that people like me volunteer here and contribute not only by quick fixes, but also debugging or bringing research ideas. The code of conduct promises a welcoming attitude and I feel this PR goes completely opposite :-(

**Comment by moorepants** [comment]
To @oscarbenjamin 

> Okay I'm sorry Jason. I understood your objection as being about the linting rule rather than the content of the PR itself and I wanted to get CI fixed so I guess I rushed merging this. We can of course consider reverting this if you are unhappy with the specific changes here.

Thank you for the apology. I do not request that the PR be reverted but I do request that we follow expected protocol and not merge if it isn't clear that there is no dissent.

> Either way we should follow up by considering what we want from this linter and we should change the config to be more selective in some way so that new lint codes don't appear unexpectedly like this.

I agree with this. We have selected a subset of flake8 (non-plugin) rules that we've all been working with for years, we should continue the explicit opt-in rules. Adding flake8-comprehensions added many new rules that I'm not sure most people are happy with and now it seems to leak in new rules without explicit adoption. We should probably tread lightly when pushing these linting rules through, given that it is known that there are strong opinions about them. SymPy will have to come to terms with moving off old paradigms into new ones, but it is a big ship to direct and we have to take baby steps.

To @maciejskorski 

> I am afraid that the objections were not clearly expressed.

I my view, you have done the right thing given that you didn't understand my objections, which was to engage me in a discussion so that we can see if we can reach some consensus.

> Please keep in mind that people like me volunteer here and contribute not only by quick fixes, but also debugging or bringing research ideas.

Most all of us are volunteers here, I've given thousands of hours to SymPy over the years. So we are generally aware that new contributors are too. I personally am grateful to every contributor and say that publicly often when I discuss SymPy.

> The code of conduct promises a welcoming attitude and I feel this PR goes a bit wrong on this aspect :-(

I apologize if you did not feel welcome. We unfortunately can't merge every  new contributor's ideas and it is also challenging to explain 15 years of history of why we don't have extensive strict linting in SymPy in a short message. There is a push from some new and existing contributors to adopt a hyper modern set of Python development practices with this 15 year old library. The intentions are good, but challenging both technically and culturally. I will try to adjust my behavior when I see I'm reviewing a new contributor's PR.

**Pull request merged**
